### PR TITLE
`init --upgrade`

### PR DESCRIPTION
**Pull request by wfleming**
This adds an `--upgrade` flag the `init` command will recognize to upgrade classic configuration.

In the upgrade case, the logic for each individual engine should be:

* Enable it if its language was supported by the classic platform & was enabled in the user's existing classic config.
* Enable it if its language was not supported by classic platform, but it meets the normal criteria for inclusion.
* Do not enable it if its language was supported by the classic platform but it was not enabled in the user's classic config.

Additionally, if upgrading a classic config, do not alter `exclude_paths`, since this key existed for classic config & should be respected.

Basically, do what we would normally do unless it contradicts existing config.

There's definitely still opportunity for me clean this up some, but I'm going to open this for review now since I think the cleaning will make the diff harder to read. cc @codeclimate/review

### CODE REVIEW

File: spec/cc/cli/init_spec.rb [hunk 1]
```diff
@@ -94,12 +82,12 @@ module CC::CLI
...
       end
 
-      describe "when a .codeclimate.yml file is already present in working directory" do
+      describe "when a platform .codeclimate.yml file is already present in working directory" do
         it "does not create a new file or overwrite the old" do
           within_temp_dir do
             filesystem.exist?(".codeclimate.yml").must_equal(false)
 
-            yaml_content_before = "This is a test yaml!"
+            yaml_content_before = "---\languages:\n  Ruby: true\n"
```
Comment (pbrisbin) [inline spec/cc/cli/init_spec.rb]: Is this supposed to be `---\nlanguages...`?

**Comment by fhwang** [comment]
`init.rb` has a lot of calls to `upgrade?` At the risk of inciting OO wankery, I wonder if it would clean things up to use a little delegation or even inheritance.

File: spec/cc/cli/init_spec.rb (hunk 1 above)
Comment (wfleming) [inline spec/cc/cli/init_spec.rb]: …Yes. I thought I had caught all the instances. Thanks.

**Comment by wfleming** [comment]
@fhwang Yep, I'm in progress on that now.

**Comment by wfleming** [comment]
Ok, this is feeling better. Another look @fhwang @codeclimate/review ?

File: lib/cc/cli/config_generator.rb [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        AUTO_EXCLUDE_PATHS.select { |path| filesystem.exist?(path) }
+      end
+
+      def post_generation_verb
+        "generated"
+      end
+
+      private
+
+      attr_reader :filesystem
```
Comment (GordonDiggs) [inline lib/cc/cli/config_generator.rb]: :lipstick: combine these 2 lines

File: lib/cc/cli/init.rb [hunk 3]
```diff
@@ -51,29 +57,17 @@ def available_configs
         all_paths.reject { |path| ['.', '..'].include?(File.basename(path)) }
       end
 
-      def engine_eligible?(engine)
-        !engine["community"] && engine["enable_regexps"].present? && has_files?(engine)
-      end
+      def engines_enabled?
+        return @engines_enabled if @engines_enabled
```
Comment (GordonDiggs) [inline lib/cc/cli/init.rb]: The one issue with this memoization is that if `@engines_enabled` is false, it won't use the memoized value. I think you can get around that by returning `unless @engines_enabled.nil?`

Comment (wfleming) [inline lib/cc/cli/init.rb]: Yep, good point.

**Comment by GordonDiggs** [comment]
LGTM
